File: packages/app/src/observability.ts
import process from "node:process";

type ObservabilityValue =
  | string
  | number
  | boolean
  | null
  | undefined
  | string[]
  | number[]
  | boolean[];

export interface ObservabilityEvent {
  layer: "runtime" | "store" | "sidecar";
  event: string;
  requestId?: string;
  method?: string;
  status?: "start" | "ok" | "error" | "cancelled" | "timeout";
  durationMs?: number;
  details?: Record<string, ObservabilityValue>;
}

const isObservabilityEnabled = (): boolean =>
  process.env.SPECFLOW_DEBUG_OBSERVABILITY === "1";

const nowIso = (): string => new Date().toISOString();

export const describeObservabilityError = (error: unknown): string => {
  if (error instanceof Error) {
    return error.message;
  }

  return String(error);
};

export const logObservabilityEvent = (event: ObservabilityEvent): void => {
  if (!isObservabilityEnabled()) {
    return;
  }

  process.stderr.write(`${JSON.stringify({
    ts: nowIso(),
    ...event
  })}\n`);
};

File: packages/tauri/src-tauri/src/sidecar_supervisor/observability.rs
use std::env;

use serde_json::Value;

pub(super) fn log_observability_event(event: &str, payload: Value) {
    if env::var("SPECFLOW_DEBUG_OBSERVABILITY")
        .map(|value| value == "1")
        .unwrap_or(false)
    {
        eprintln!(
            "{}",
            serde_json::json!({
                "layer": "desktop-bridge",
                "event": event,
                "payload": payload
            })
        );
    }
}
